File: wisp-swift/Sources/WispAI/RPC.swift
import Foundation

// MARK: - Protocol types

public struct RpcCommand: Decodable {
    public let type: String
    public let id: String?
    public let message: String?
    public let provider: String?
    public let modelId: String?
    public let systemPrompt: String?
}

public struct RpcResponse: Encodable {
    public let id: String?
    public let type: String
    public let command: String
    public let success: Bool
    public let data: AnyEncodable?
    public let error: String?

    public init(id: String?, command: String, data: (any Encodable)? = nil, error: String? = nil) {
        self.id = id
        self.type = "response"
        self.command = command
        self.success = error == nil
        self.data = data.map(AnyEncodable.init)
        self.error = error
    }
}

public struct RpcSessionState: Encodable {
    public let model: Model
    public let isStreaming: Bool
    public let sessionId: String
    public let messageCount: Int
}

struct RpcEvent: Encodable {
    let type: EventType
    let index: Int?
    let delta: String?
    let content: String?
    let part: ContentPart?
    let message: AssistantMessage?

    init(from e: AssistantMessageEvent) {
        type = e.type
        index = e.index != 0 ? e.index : nil
        delta = e.delta.isEmpty ? nil : e.delta
        content = e.content.isEmpty ? nil : e.content
        part = e.part
        message = e.message
    }
}

// MARK: - Type-erased encodable

public struct AnyEncodable: Encodable {
    private let _encode: (Encoder) throws -> Void
    public init(_ value: any Encodable) { _encode = { try value.encode(to: $0) } }
    public func encode(to encoder: Encoder) throws { try _encode(encoder) }
}

// MARK: - Session actor

actor RPCSession {
    var messages: [ContextMessage] = []
    var systemPrompt: String
    var model: Model
    var options: StreamOptions
    var tools: [Tool]
    var handler: ToolHandler?
    var isStreaming: Bool = false
    var streamTask: Task<Void, Never>?
    var id: String
    let config: Config

    init(config: Config, model: Model, options: StreamOptions, tools: [Tool],
         handler: ToolHandler?, systemPrompt: String) {
        self.config = config; self.model = model; self.options = options
        self.tools = tools; self.handler = handler; self.systemPrompt = systemPrompt
        self.id = UUID().uuidString
    }

    func abort() { streamTask?.cancel(); streamTask = nil; isStreaming = false }

    func newSession() {
        abort(); messages = []; systemPrompt = ""; id = UUID().uuidString
    }

    func startStreaming(task: Task<Void, Never>) {
        streamTask = task; isStreaming = true
    }

    func finishStreaming() { streamTask = nil; isStreaming = false }

    func appendMessages(_ msgs: [ContextMessage]) { messages.append(contentsOf: msgs) }

    func setModel(_ m: Model, opts: StreamOptions) { model = m; options = opts }

    func setSystemPrompt(_ s: String) { systemPrompt = s }

    func snapshot() -> (model: Model, options: StreamOptions, systemPrompt: String,
                        messages: [ContextMessage], tools: [Tool], handler: ToolHandler?,
                        isStreaming: Bool, id: String) {
        (model, options, systemPrompt, messages, tools, handler, isStreaming, self.id)
    }
}

// MARK: - Output writer actor

actor OutputWriter {
    func write(_ data: Data) {
        var buf = data
        buf.append(contentsOf: [UInt8(ascii: "\n")])
        FileHandle.standardOutput.write(buf)
    }

    func writeEncodable(_ value: some Encodable) {
        guard let data = try? JSONEncoder().encode(value) else { return }
        write(data)
    }
}

// MARK: - RPC server

public func runRPC(
    config: Config,
    model: Model,
    options: StreamOptions,
    tools: [Tool],
    handler: ToolHandler?,
    initialSystemPrompt: String
) async {
    let session = RPCSession(config: config, model: model, options: options,
                             tools: tools, handler: handler, systemPrompt: initialSystemPrompt)
    let out = OutputWriter()

    func respond(id: String?, command: String, data: (any Encodable)? = nil, error: String? = nil) async {
        await out.writeEncodable(RpcResponse(id: id, command: command, data: data, error: error))
    }

    do {
        for try await line in FileHandle.standardInput.bytes.lines {
            let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !trimmed.isEmpty,
                  let cmd = try? JSONDecoder().decode(RpcCommand.self, from: Data(trimmed.utf8))
            else { continue }

            switch cmd.type {

            case "prompt", "follow_up":
                let snap = await session.snapshot()
                if snap.isStreaming {
                    await respond(id: cmd.id, command: cmd.type, error: "already streaming; send abort first")
                    continue
                }
                let userMsg = ContextMessage.user(cmd.message ?? "")
                await session.appendMessages([userMsg])
                let snap2 = await session.snapshot()
                let ctx = Context(systemPrompt: snap2.systemPrompt, messages: snap2.messages, tools: snap2.tools)
                let resolvedOpts = snap2.options

                await respond(id: cmd.id, command: cmd.type)

                let task = Task {
                    defer { Task { await session.finishStreaming() } }

                    let agentOpts = AgentOptions(onStep: { added in
                        Task { await session.appendMessages(added) }
                    })

                    let noopHandler: ToolHandler = { _, name, _ in "error: no handler for tool \(name)" }
                    let h = snap2.handler ?? noopHandler

                    for try await event in runAgent(config: snap2.config, model: snap2.model,
                                                    context: ctx, options: resolvedOpts,
                                                    handler: h, agentOptions: agentOpts) {
                        await out.writeEncodable(RpcEvent(from: event))
                        if event.type == .done, let msg = event.message {
                            let lastRole = await session.snapshot().messages.last?.role
                            if lastRole != "assistant" {
                                await session.appendMessages([.assistant(msg)])
                            }
                        }
                    }
                }
                await session.startStreaming(task: task)

            case "abort":
                await session.abort()
                await respond(id: cmd.id, command: "abort")

            case "new_session":
                await session.newSession()
                await respond(id: cmd.id, command: "new_session")

            case "get_state":
                let snap = await session.snapshot()
                let state = RpcSessionState(model: snap.model, isStreaming: snap.isStreaming,
                                            sessionId: snap.id, messageCount: snap.messages.count)
                await respond(id: cmd.id, command: "get_state", data: state)

            case "get_messages":
                let msgs = await session.snapshot().messages
                await respond(id: cmd.id, command: "get_messages", data: ["messages": msgs])

            case "set_model":
                let registry = config.registry
                let pid = cmd.provider ?? ""
                let mid = cmd.modelId ?? ""
                let found = registry.find(provider: pid, id: mid) ?? registry.findByID(mid)
                if let m = found {
                    let newOpts = config.resolve(nil, for: m)
                    await session.setModel(m, opts: newOpts)
                    await respond(id: cmd.id, command: "set_model", data: m)
                } else {
                    await respond(id: cmd.id, command: "set_model", error: "model \"\(mid)\" not found")
                }

            case "get_available_models":
                let models = config.registry.all()
                await respond(id: cmd.id, command: "get_available_models", data: ["models": models])

            case "set_system_prompt":
                await session.setSystemPrompt(cmd.systemPrompt ?? "")
                await respond(id: cmd.id, command: "set_system_prompt")

            default:
                await respond(id: cmd.id, command: cmd.type, error: "unknown command: \(cmd.type)")
            }
        }
    } catch {
        // stdin closed or error — exit cleanly
    }
}

File: wisp/wispai/rpc.go
package wispai

import (
	"bufio"
	"context"
	"crypto/rand"
	"encoding/json"
	"fmt"
	"io"
	"sync"
)

// --- Protocol types (stdin → commands, stdout → responses + events) ---

// RpcCommand is a command sent by the UI to the agent via stdin (JSON Lines).
type RpcCommand struct {
	Type         string `json:"type"`
	ID           string `json:"id,omitempty"`
	Message      string `json:"message,omitempty"`
	Provider     string `json:"provider,omitempty"`
	ModelID      string `json:"modelId,omitempty"`
	SystemPrompt string `json:"systemPrompt,omitempty"`
}

// RpcResponse is sent to stdout in reply to a command.
type RpcResponse struct {
	ID      string `json:"id,omitempty"`
	Type    string `json:"type"` // always "response"
	Command string `json:"command"`
	Success bool   `json:"success"`
	Data    any    `json:"data,omitempty"`
	Error   string `json:"error,omitempty"`
}

// RpcSessionState is returned by get_state.
type RpcSessionState struct {
	Model        *Model `json:"model"`
	IsStreaming  bool   `json:"isStreaming"`
	SessionID    string `json:"sessionId"`
	MessageCount int    `json:"messageCount"`
}

// rpcEvent is the JSON-serializable form of AssistantMessageEvent,
// emitted to stdout during streaming.
type rpcEvent struct {
	Type    EventType         `json:"type"`
	Index   int               `json:"index,omitempty"`
	Delta   string            `json:"delta,omitempty"`
	Content string            `json:"content,omitempty"`
	Part    *ContentPart      `json:"part,omitempty"`
	Message *AssistantMessage `json:"message,omitempty"`
}

// --- RPC session state ---

type rpcSession struct {
	mu           sync.Mutex
	id           string
	messages     []ContextMessage
	systemPrompt string
	model        Model
	opts         StreamOptions
	tools        []Tool
	handler      ToolHandler
	cancelStream context.CancelFunc
	isStreaming  bool
}

// RunRPC reads JSON commands from in and writes JSON events/responses to out.
// Runs until in is closed or ctx is cancelled.
//
// Supported commands (stdin):
//
//	prompt / follow_up  — send a message, streams events back
//	abort               — cancel the current stream
//	new_session         — clear history and start fresh
//	get_state           — return current session state
//	get_messages        — return full message history
//	set_model           — change the active model
//	get_available_models — list all models from config
//	set_system_prompt   — set the system prompt for the session
//
// Every command produces a { type:"response", command, success, data?, error? }
// response on stdout. Streaming events are emitted as they arrive.
func RunRPC(
	ctx context.Context,
	config *Config,
	model Model,
	opts StreamOptions,
	tools []Tool,
	handler ToolHandler,
	initialSystemPrompt string,
	in io.Reader,
	out io.Writer,
) {
	s := &rpcSession{
		id:           rpcNewID(),
		model:        model,
		opts:         opts,
		tools:        tools,
		handler:      handler,
		systemPrompt: initialSystemPrompt,
	}

	var outMu sync.Mutex
	write := func(v any) {
		b, err := json.Marshal(v)
		if err != nil {
			return
		}
		outMu.Lock()
		out.Write(append(b, '\n')) //nolint
		outMu.Unlock()
	}

	respond := func(id, command string, data any, errMsg string) {
		r := RpcResponse{
			ID:      id,
			Type:    "response",
			Command: command,
			Success: errMsg == "",
		}
		if errMsg != "" {
			r.Error = errMsg
		} else {
			r.Data = data
		}
		write(r)
	}

	scanner := bufio.NewScanner(in)
	scanner.Buffer(make([]byte, 4*1024*1024), 4*1024*1024)

	for scanner.Scan() {
		if ctx.Err() != nil {
			return
		}
		line := scanner.Bytes()
		if len(line) == 0 {
			continue
		}
		var cmd RpcCommand
		if err := json.Unmarshal(line, &cmd); err != nil {
			continue
		}

		switch cmd.Type {

		case "prompt", "follow_up":
			s.mu.Lock()
			if s.isStreaming {
				s.mu.Unlock()
				respond(cmd.ID, cmd.Type, nil, "already streaming; send abort first")
				continue
			}
			s.messages = append(s.messages, UserMsg(cmd.Message))
			msgCtx := &Context{
				SystemPrompt: s.systemPrompt,
				Messages:     append([]ContextMessage{}, s.messages...),
				Tools:        s.tools,
			}
			streamCtx, cancel := context.WithCancel(ctx)
			s.cancelStream = cancel
			s.isStreaming = true
			currentModel := s.model
			currentOpts := s.opts
			s.mu.Unlock()

			respond(cmd.ID, cmd.Type, nil, "")

			go func(id string) {
				defer func() {
					cancel()
					s.mu.Lock()
					s.isStreaming = false
					s.cancelStream = nil
					s.mu.Unlock()
				}()

				agentOpts := &AgentOptions{
					OnStep: func(added []ContextMessage) {
						s.mu.Lock()
						s.messages = append(s.messages, added...)
						s.mu.Unlock()
					},
				}

				for e := range RunAgent(streamCtx, config, currentModel, msgCtx, &currentOpts, s.handler, agentOpts) {
					write(rpcEvent{
						Type:    e.Type,
						Index:   e.Index,
						Delta:   e.Delta,
						Content: e.Content,
						Part:    e.Part,
						Message: e.Message,
					})
					// For text-only responses (no tool calls), RunAgent emits
					// EventDone without calling OnStep, so append the message here.
					if e.Type == EventDone && e.Message != nil {
						s.mu.Lock()
						// Only append if the message isn't already in history
						// (OnStep handles the tool-use path).
						last := len(s.messages) - 1
						alreadyStored := last >= 0 && s.messages[last].Role == "assistant"
						if !alreadyStored {
							s.messages = append(s.messages, AssistantMsg(e.Message))
						}
						s.mu.Unlock()
					}
				}
			}(cmd.ID)

		case "abort":
			s.mu.Lock()
			if s.cancelStream != nil {
				s.cancelStream()
			}
			s.mu.Unlock()
			respond(cmd.ID, "abort", nil, "")

		case "new_session":
			s.mu.Lock()
			if s.cancelStream != nil {
				s.cancelStream()
			}
			s.messages = nil
			s.systemPrompt = ""
			s.id = rpcNewID()
			s.mu.Unlock()
			respond(cmd.ID, "new_session", nil, "")

		case "get_state":
			s.mu.Lock()
			m := s.model
			state := RpcSessionState{
				Model:        &m,
				IsStreaming:  s.isStreaming,
				SessionID:    s.id,
				MessageCount: len(s.messages),
			}
			s.mu.Unlock()
			respond(cmd.ID, "get_state", state, "")

		case "get_messages":
			s.mu.Lock()
			msgs := append([]ContextMessage{}, s.messages...)
			s.mu.Unlock()
			respond(cmd.ID, "get_messages", map[string]any{"messages": msgs}, "")

		case "set_model":
			registry := config.Registry()
			m, ok := registry.Find(cmd.Provider, cmd.ModelID)
			if !ok {
				m, ok = registry.FindByID(cmd.ModelID)
			}
			if !ok {
				respond(cmd.ID, "set_model", nil, fmt.Sprintf("model %q not found", cmd.ModelID))
			} else {
				newOpts := config.Resolve(nil, m)
				s.mu.Lock()
				s.model = m
				s.opts = newOpts
				s.mu.Unlock()
				respond(cmd.ID, "set_model", m, "")
			}

		case "get_available_models":
			models := config.Registry().All()
			respond(cmd.ID, "get_available_models", map[string]any{"models": models}, "")

		case "set_system_prompt":
			s.mu.Lock()
			s.systemPrompt = cmd.SystemPrompt
			s.mu.Unlock()
			respond(cmd.ID, "set_system_prompt", nil, "")

		default:
			respond(cmd.ID, cmd.Type, nil, fmt.Sprintf("unknown command: %s", cmd.Type))
		}
	}
}

func rpcNewID() string {
	b := make([]byte, 16)
	rand.Read(b)
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:])
}
